File: step-01-basics/language-warmup/sol.py
"""
problem :
link    :
pattern :
brute   : O( ) time, O( ) space
optimal : O( ) time, O( ) space
"""
from typing import List, Optional
from collections import defaultdict, deque, Counter
from heapq import heappush, heappop, heapify
from bisect import bisect_left, bisect_right


def solve():
    str = input("Enter a string: ")
    num = int(input("Enter an int: "))
    numlist = list(map(int,input().split()))
    print(f'str: {str}, num: {num}, numlist: {numlist}')

    marks = int(input("Enter marks: "))
    if(marks<40):
        print("Fail")
    else:
        print("OK")

    match marks:
        case 40:
            print("Damn just passed")

    for i in range(0,5):
        print(f'{i} ')

    i = 0
    while(i<5):
        print(f'{i} ')
        i+=1

    def try_swap(a, b):
        a, b = b, a

    def mutate(lst):
        lst.append(99)

    a, b = 10, 20
    try_swap(a,b)
    print(f'a: {a}, b: {b}')
    lst = [1,2,3,4]
    mutate(lst)
    print(f'{lst}')


solve()

File: step-01-basics/language-warmup/sol.cpp
// problem :
// link    :
// pattern :
// brute   : O( ) time, O( ) space
// better  : O( ) time, O( ) space
// optimal : O( ) time, O( ) space

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <queue>
#include <stack>
#include <climits>
#include <cmath>
using namespace std;

void swapByVal(int a, int b)
{
    int temp = b;
    b = a;
    a = temp;
}

void swapByRef(int &a, int &b)
{
    int temp = a;
    a = b;
    b = temp;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    cout << "Enter an int, double and a string: " << endl;
    int num;
    double dob;
    string str;
    cin >> num >> dob >> str;
    cout << "Int entered is: " << num << ", Double entered is: " << dob << "String entered is: str" << endl;

    cout << sizeof(int) << " | " << sizeof(double) << " | " << sizeof(long long) << " | " << sizeof(char) << endl;
    cout << numeric_limits<int>::max() << "-" << numeric_limits<int>::min() << endl;
    cout << numeric_limits<long long>::max() << "-" << numeric_limits<long long>::min() << endl;
    cout << numeric_limits<double>::max() << "-" << numeric_limits<double>::min() << endl;
    cout << numeric_limits<char>::max() << "-" << numeric_limits<char>::min() << endl;

    int marks;
    cout << "Enter marks: " << endl;
    cin >> marks;
    if (marks < 40)
    {
        cout << "Fail" << endl;
    }
    else if (marks < 70)
    {
        cout << "Pass but do better" << endl;
    }
    else
    {
        cout << "Pass" << endl;
    }

    array<int, 4> arr = {0, 1, 2, 3};
    vector<int> vec = {0, 1, 2, 3, 4, 5};

    char cStr[] = "Hello";
    string text = "Hey There";

    cout << "Array length: " << arr.size() << endl;
    cout << "Vector length: " << vec.size() << endl;
    cout << "C-String length: " << strlen(cStr) << endl;
    cout << "String length: " << text.length() << endl;

    for (int i : vec)
    {
        cout << "Vec " << i << endl;
    }

    for (char c : text)
    {
        cout << "Text " << c << endl;
    }

    for (int i = 0; i < strlen(cStr); i++)
    {
        cout << "cStr " << cStr[i] << endl;
    }

    for (int i : arr)
    {
        cout << "Arr " << i << endl;
    }

    for (int i = 0; i < 10; i++)
    {
        cout << i << " ";
    }

    cout << "\n";

    int i = 0;
    while (i < 10)
    {
        cout << i << " ";
        i++;
    }

    cout << "\n";

    int a = 10;
    int b = 20;
    swapByVal(a, b);
    cout << "a: " << a << " b: " << b << endl;
    swapByRef(a, b);
    cout << "a: " << a << " b: " << b << endl;

    return 0;
}
